Lib/pathlib.py

```diff
@@ -1385,6 +1386,121 @@ def expanduser(self):
+        will be called with one argument, an OSError instance.  It can
+        report the error to continue with the walk, or raise the exception
+        to abort the walk.  Note that the filename is available as the
+        filename attribute of the exception object.
+
+        By default, Path.walk does not follow symbolic links to subdirectories
+        on systems that support them.  In order to get this functionality, set
+        the optional argument 'follow_symlinks' to true. Unlike os.walk,
+        Path.walk only adds symbolic links to dirnames if follow_symlinks=True. 
+
+        Caution: if self is a relative Path, don't change the
+        current working directory between resumptions of walk. walk never
+        changes the current directory, and assumes that the caller doesn't
+        either.
+        
+        Caution: Unlike os.walk, Path.walk assumes the directories have not
+        been modified between its resumptions. I.e. If a directory from
+        dirnames has been replaced with a symlink and follow_symlinks=False,
+        walk will still try to descend into it. To prevent such behavior,
+        remove directories from dirnames if they have been modified and you
+        do not want Path.walk to descend into them anymore.
+
+        Example:
+
+        from pathlib import Path
+        for root, dirs, files in Path().walk(on_error=print):
+            print(
+                root,
+                "consumes",
+                sum((root / file).stat().st_size for file in files),
+                "bytes in",
+                len(files),
+                "non-directory files"
+            )
+            # don't visit __pycache__ directories
+            if '__pycache__' in dirs:
+                dirs.remove('__pycache__')
+        """
+        sys.audit("pathlib.Path.walk", self, on_error, follow_symlinks)
+        return self._walk(top_down, on_error, follow_symlinks)
+
+    def _walk(self, top_down, on_error, follow_symlinks):
+        dirs = []
+        nondirs = []
+
+        # We may not have read permission for self, in which case we can't
+        # get a list of the files the directory contains. os.walk
+        # always suppressed the exception then, rather than blow up for a
+        # minor reason when (say) a thousand readable directories are still
+        # left to visit. That logic is copied here.
+        try:
+            scandir_it = self._scandir()
+        except OSError as error:
+            if on_error is not None:
+                on_error(error)
+            return
+
+        with scandir_it:
+            for entry in scandir_it:
+                try:
+                    is_dir = entry.is_dir(follow_symlinks=follow_symlinks)
+                except OSError:
+                    # If is_dir() raises an OSError, consider that the entry
+                    # is not a directory, same behavior as os.path.isdir()
+                    is_dir = False
```

Comment:
Nit: this is self-explanatory and doesn't require a comment IMO

Reply:
I partially disagree. I will leave the last statement if that's fine with you